「Webセキュリティ」のプログラムを作成(修正)しました

a-kuroki-gs opened the pull request:
フィヨルドブートキャンプの[Webセキュリティ](https://bootcamp.fjord.jp/practices/231)の課題である、
様々な脆弱性を抱えた呪われたブログアプリの脆弱性を取り除き、すべてのテストをパスするように修正しました。

a-kuroki-gs committed:
READMEに沿ってセットアップを行う

a-kuroki-gs committed:
'検索キーワードに ') OR secret = TRUE -- を入力すると秘密のブログが見えちゃう！'を修正

app/models/blog.rbでプレースホルダを用いてSQLインジェクション対策を行う

a-kuroki-gs committed:
'他の人のブログ記事を編集できちゃう！'を修正

edit・update・destroyアクションで自分のブログから取って来るメソッドを定義するように修正
見つからなければ例外を捕捉し、フラッシュメッセージを表示

a-kuroki-gs committed:
'他の人の秘密のブログにアクセスできちゃう！'を修正

コントローラのshowアクションを編集し、秘密のブログかどうか・自分のブログかどうかをチェックする
他人の秘密のブログであれば表示できないようにする

a-kuroki-gs committed:
'ブログ本文に <script>alert(XSS)</script> を書き込むとバッチリJavaScriptが実行されちゃう！'を修正

sanitizeとhtml_escapeを使ってエスケープする

a-kuroki-gs committed:
'HTMLをいじって <input type=hidden name=blog[random_eyecatch] value=1> を追加すると誰でも有料機能が使えちゃう！'を修正

ストロングパラメータで条件分岐を行い制御するように修正

a-kuroki-gs committed:
'APIのレスポンスを覗いたら個人情報がわかっちゃう！'を修正

JSONを返すところでidとnicknameのみ返すようにした
他のものを返すと、それがAPIレスポンスとして返ってしまうため

a-kuroki-gs committed:
テストでパスしてなかったものをパスするようにする

ユーザーの権限周りのところで、例外を受け取ってリダイレクトさせてしまっていた
例外を発生させたままにすることがテストで求められていたので、修正

cafedomancer commented on yarn.lock:
余分な変更が加えられているように見えます。課題をクリアするのに必要最低限の変更のみを加えてください。

a-kuroki-gs replied:
不要な変更を削除し、コミットしております。

cafedomancer commented on package-lock.json:
余分な変更が加えられているように見えます。課題をクリアするのに必要最低限の変更のみを加えてください。

a-kuroki-gs replied:
こちらも不要な変更を削除し、コミットしております。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -1,15 +1,21 @@
 
-  def show; end
+  def show
+    raise ActiveRecord::RecordNotFound if @blog.secret? && @blog.user != current_user
```
`ActiveRecord::RecordNotFound` は自分で送出するのではなく、Rails の API を使った上で送出されるようにしてください (そのような実装を考えてください) 。

a-kuroki-gs replied:
`find_by`と`find_by!`を使って、
`raise`しなくても`ActiveRecord::RecordNotFound`が発生するよう修正しました。

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,6 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    html_escape(blog.content).gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
```
`rubocop:disable` がなくても問題なく動くような実装にできないか考えてみてください。

a-kuroki-gs replied:
`html_safe`でエラーが出ていたので、`sanitize`を使ってエスケープするよう修正しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -47,7 +53,20 @@ def set_blog
+    @blog = current_user.blogs.find(params[:id])
+  end
+
+  def sanitize_blog_content
+    tags = %w[a acronym b strong i em li ul ol h1 h2 h3 h4 h5 h6 blockquote br cite sub sup ins p]
+    blog_params[:content] = sanitize(blog_params[:content], tags:, attributes: %w[href title])
+  end
```
この処理は本当に意図どおりに動いていますか？ `blog_params` は変数ではなくメソッドで、呼び出しのたびに新しい値が返されるため、`blog_params[:content]` のような書き換えを行っても意味がないのでは、と思ったからです。そもそもですが、この処理はなんのために必要ですか？

a-kuroki-gs replied:
> `blog_params` は変数ではなくメソッドで、呼び出しのたびに新しい値が返されるため、`blog_params[:content]` のような書き換えを行っても意味がないのでは、と思ったからです。

すみません。こちら私の勉強不足・理解不足でした。
入力された`content`に対して`sanitize`メソッドを使ってエスケープ処理を行っているつもりになっていました。

こちら不要でしたので削除しております。
また、エスケープ処理は `app/helpers/blogs_helper.rb`で実装していたので、追加の実装はありません。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -47,7 +53,20 @@ def set_blog
+
   def blog_params
-    params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    if current_user.premium?
+      params.require(:blog).permit(:title, :content, :secret, :random_eyecatch)
+    else
+      params.require(:blog).permit(:title, :content, :secret)
+    end
```
`:random_eyecatch` の指定以外の部分が分岐の間で重複していそうです。

a-kuroki-gs replied:
こちらご指摘の内容を反映できているかやや不安な部分ではありますが、
`:random_eyecatch` の指定以外の部分を切り出して修正しております。

修正が不十分でしたらコメントいただけますと幸いです。

a-kuroki-gs committed:
不要な差分を削除する

push後の削除方法がわからなかったので、差分が起きないように修正して再度commit→push

a-kuroki-gs committed:
rubocopを無理矢理通していた部分の修正

html_safeでエラーが出ていたので、sanitizeを使ってエスケープするよう修正

a-kuroki-gs committed:
分岐の間で重複していた:random_eyecatchの指定以外の部分を切り出して修正

a-kuroki-gs committed:
ActiveRecord::RecordNotFoundをraiseしなくても発生するようにロジックを修正

a-kuroki-gs committed:
コントローラでsanitizeを行う処理を削除

blog_paramsの中身にsanitizeメソッドを使っていたが、blog_paramsは呼び出しのたびに新しい値が返されるので書き換えても意味がないため
app/helpers/blogs_helper.rbで制御できていたので追加の処理はなし

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,16 @@
 
-  def show; end
+  def show
+    @blog = Blog.find_by(id: params[:id], secret: false) ||
+            Blog.find_by!(id: params[:id], secret: true, user: current_user)
```
これだと `Blog` の検索が最大で 2 回行われてしまいそうです。不要なクエリはなるべく発生させたくないので、`Blog` の検索を 1 回で済ませられるように、条件を整理してクエリを考え直してみてください。

a-kuroki-gs replied:
>条件を整理してクエリを考え直してみてください。

こちら考え直したところ、`user: current_user`のときは`secret: true`の評価がそもそも不要でした。

> これだと `Blog` の検索が最大で 2 回行われてしまいそうです。

すみません。こちらはDBに取得しにいくのが最大で２回という認識でよろしいですか？
検索が２回、の基準が間違っていれば教えていただきたいです。

今回は、`where`と`find`を使って最低限のクエリで取得するように修正しました。
この修正でよろしいでしょうか？

cafedomancer replied:
はい、よいですー。改善できる点があるので別途コメントします。

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,6 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    sanitize(html_escape(blog.content).gsub("\n", '<br>'))
```
`html_escape` した結果を渡しているので問題にはなりませんが、`sanitize` すると `<br>` 以外のタグも有効になる可能性があるので、別の方法を考えたほうがよさそうです。

a-kuroki-gs replied:
ここは改行を反映させたい処理でしたので、`sanitize`は使うべきではありませんでした。

調べたところ`simple_format`メソッドを使うと改行を反映できるということがわかりましたので、
そのように修正しました。

a-kuroki-gs committed:
whereメソッドとfindメソッドを使って、最低限のクエリでblogを取得できるよう修正

クエリを２回実行する可能性があったところを、１回で済むように修正するため
whereメソッドを使って、見られるblogをあらかじめ絞り込む
それからparams[:id]を使ってblogを取得することで例外発生にも対応する

a-kuroki-gs committed:
改行を反映させるところでsanitize を使ってしまっていたのをsimple_formatに修正

sanitize すると <br> 以外のタグも有効になる可能性があるため
simple_formatを利用すると改行のみ影響がある

cafedomancer commented on app/helpers/blogs_helper.rb:
```diff
@@ -2,6 +2,6 @@
 module BlogsHelper
   def format_content(blog)
-    blog.content.gsub("\n", '<br>').html_safe # rubocop:disable Rails/OutputSafety
+    simple_format(html_escape(blog.content).gsub("\n", '<br>'))
```
`simple_format` を使うなら `gsub` による変換はいらないと思うので確認してみてください。

a-kuroki-gs replied:
`simple_format`・`html_escape`・`gsub`について調べ直しました。
`html_escape`でHTMLタグをエスケープ→`simple_format`で改行に`<br />`を付与
という流れが正しいかと思いますので、そのように修正しました。

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,15 @@
 
-  def show; end
+  def show
+    @blog = Blog.where('secret = ? or user_id = ?', false, current_user&.id).find(params[:id])
```
この方法でいくなら SQL のキーワードは大文字で書いておくようにしましょう。

```suggestion
    @blog = Blog.where('secret = ? OR user_id = ?', false, current_user&.id).find(params[:id])
```

ところで、Active Record の機能を使うと別の書き方もできそうです。いくつかヒントを書いておくのでトライしてみてください。

- 引数に文字列ではなくハッシュを使う
- 関連の指定に ID ではなくオブジェクトを使う
- OR を Active Record の機能を使って表現する

a-kuroki-gs replied:
ご指摘いただいたように、Active Record の機能を使って修正しております。

こちら私の理解不足もあったかもしれません。
`Blog.where(...)`のところでクエリが発生すると考えており、`Blog.where(...)`と重ねて書くこと、つまり今回で言うと
`Blog.where(...).or(Blog.where(...))`のようにする修正は行うべきではないと認識してしまっておりました。

`where`はクエリを発行するがDBへのアクセスは行わず、アクセスを行うのは`find`・`find_by`などという理解でよいのでしょうか？

cafedomancer replied:
> `where`はクエリを発行するがDBへのアクセスは行わず、アクセスを行うのは`find`・`find_by`などという理解でよいのでしょうか？

`find` メソッドと `find_by` メソッドについてはその認識でよいと思います。`where` メソッドは、データが実際に必要になるタイミングまでクエリの発行を遅延するメソッドです。クエリを発行しないわけではありません。例えば `rails console` で `Blog.where(secret: true)` のようなコードを実行してみると、`where` メソッドでもクエリの発行が行われていることが分かります。これは、実行結果がコンソールですぐに必要になったからです。このように、クエリの発行を遅延するメソッドとそうでないメソッドがあることを覚えておいてください。

a-kuroki-gs replied:
ご回答ありがとうございます。
クエリの発行の流れが理解できていなかったので、大変勉強になりました。
今回の場合で言うと、
「`where`→まだクエリを発行しない(遅らせている)→`find`→ここでタイミングがきたからここで発行する」という流れだったのですね。

また、提出物にも記載しているように、コンソールを使って試してみるという考えがすっかり抜けていたので、良い気付きになりました。ありがとうございます。

a-kuroki-gs committed:
改行を反映させる部分で、gsub による変換を削除

simple_formatで<br />を付与してくれるので、不要であったため

a-kuroki-gs committed:
blogを探してくるSQLの部分を、Active Record の機能を使ってわかりやすく書き換える

引数は文字列ではなくハッシュにする
userのIDを指定していたが、IDではなくオブジェクトで指定
whereとorを使って条件を記述する

cafedomancer commented on app/controllers/blogs_controller.rb:
```diff
@@ -3,13 +3,15 @@
 
-  def show; end
+  def show
+    @blog = Blog.where(secret: false).or(Blog.where(user: current_user)).find(params[:id])
```
`where(secret: false)` の部分は既存の実装が使えるはずなので、既存のコードを読んでみてください。

a-kuroki-gs replied:
`app/models/blog.rb`に定義されていた
`scope :published, -> { where('secret = FALSE') }`の内容が使えそうだと判断しました。

`Blog.where(secret: false)`を`Blog.published`に修正しました。

a-kuroki-gs committed:
showアクションの探すblogの条件を、既存で定義しているpublishを用いるよう変更
